File: internal/licensing/waivers.go
package licensing

import (
	"encoding/json"
	"fmt"
	"net/http"
	neturl "net/url"
)

// Waiver is an admin's decision (Starter+, via the dashboard) to accept a
// specific finding rather than have it block merges — matched by
// category+resource+file within one repo, not by line number (line shifts
// when unrelated code above it changes; requiring an exact line match
// would make a waiver go stale on the next unrelated edit).
type Waiver struct {
	Category      string `json:"category"`
	Resource      string `json:"resource"`
	FilePath      string `json:"file"`
	Justification string `json:"justification"`
}

// GetWaivers fetches every active (non-expired) waiver configured for
// repoFullName. Returns an empty slice (not an error) when none exist —
// the normal state for a repo nobody has waived anything on yet.
func (c *Client) GetWaivers(repoFullName string) ([]Waiver, error) {
	url := c.APIBase + "/v1/waivers?repo=" + neturl.QueryEscape(repoFullName)
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("building waivers request: %w", err)
	}
	req.Header.Set("Authorization", "Bearer "+c.APIKey)

	resp, err := c.HTTP.Do(req)
	if err != nil {
		return nil, fmt.Errorf("reaching licensing service: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusUnauthorized {
		return nil, fmt.Errorf("invalid or revoked API key")
	}
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("licensing service returned %s", resp.Status)
	}

	var waivers []Waiver
	if err := json.NewDecoder(resp.Body).Decode(&waivers); err != nil {
		return nil, fmt.Errorf("parsing waivers response: %w", err)
	}
	return waivers, nil
}

File: src/tfpdf/licensing/waivers.py
"""Finding-specific waivers granted by an administrator in the hosted dashboard."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(slots=True, frozen=True)
class Waiver:
    """An administrator's decision to accept a finding instead of blocking merges, available on
    Starter and above.
    """

    category: str = ""
    resource: str = ""
    file_path: str = ""
    justification: str = ""


def waivers_from_json(document: Any) -> list[Waiver]:
    """Decode waivers; an empty list is normal when none have been granted."""
    if not isinstance(document, list):
        return []
    out: list[Waiver] = []
    for raw in document:
        if not isinstance(raw, dict):
            continue
        out.append(
            Waiver(
                category=str(raw.get("category", "")),
                resource=str(raw.get("resource", "")),
                file_path=str(raw.get("file", "")),
                justification=str(raw.get("justification", "")),
            )
        )
    return out
